### ISSUE TITLE
Ability to extend or shape the graphql queries generated by codegen without writing completely new queries/mutations

### ISSUE THREAD

**Issue by hisham** [aws-amplify/amplify-cli#773]
**Is your feature request related to a problem? Please describe.**
Codgen is great for getting started, but there's often cases where we need to customize the queries and create new custom queries. Cases where certain fields would be too expensive to resolve, or more nesting depth is required, or including optional variables in nested fields (codegen I believe defaults only retrieving required variables for nested objects). If there are field resolvers that are slow, or highly relational data, then the default queries codegen generates could degrade performance unnecessarily. One can argue Apollo has a great caching layer so what's wrong with asking for as much info as possible in a query since it's a one time hit but you still run into cases like mutations where the query that is run after a mutation could be more than what you want. 

**Describe the solution you'd like**
The current solution is to write your own custom queries, disregarding what codgen produces from the GraphQL schema. I think a better solution at minimum is for the codgen code to use fragments when generating queries (see https://www.apollographql.com/docs/react/advanced/fragments.html). And maybe then during runtime, in different parts of the code if a query needs to be customized, we can just pass in a different fragment instead completely writing the new query. 

Another possible solution is some kind of dynamic query generator that works off of a base query codegen generates but then you pass it fields to include or omit (kind of like lodash's pick or omit functionality (https://lodash.com/docs/4.17.11#omit)) and it gives you the customized query back as requested. 

**Describe alternatives you've considered**
The current solution is to write your own custom queries and include them in the dir that codgen scans to generate types and APIs. This works fine but these custom queries have to then be manually updated when the GraphQL schema changes. Also you can run into cases where codgen generates multiple variations of types that all refer to the same object but the compiler or transpiler doesn't recognize them as so and so then you have to re-cast to unknown and then back to the type you want. 

**Additional context**
I am new to graphql and still learning. Above are my suggestions based on what I know and ran into developing a graphql client but the community may have an already established way to to do this. There might be things to learn from the sql world as well since this is very similar, codegen is essentially generating "select *" sql scripts for us when all I need is something similar to Rails' Active Record "select" method to pass in the variables I need (https://guides.rubyonrails.org/active_record_querying.html#selecting-specific-fields).

**Comment by regenrek** [comment]
Hey @hisham did you find any workaround for this scenario?

**Comment by hisham** [comment]
@regenrek not really. I'm just writing custom queries for now. See https://github.com/aws-amplify/amplify-cli/issues/745#issuecomment-456861984
